src/xrpld/app/tx/detail/LoanPay.cpp

```diff
@@ -0,0 +1,545 @@
+        tx.isFlag(tfLoanFullPayment) ? loanMakeFullPayment(
+                                           asset,
+                                           view,
+                                           loanSle,
+                                           brokerSle,
+                                           amount,
+                                           tx.isFlag(tfLoanOverpayment),
+                                           j_)
+                                     : loanMakePayment(
+                                           asset,
+                                           view,
+                                           loanSle,
+                                           brokerSle,
+                                           amount,
+                                           tx.isFlag(tfLoanOverpayment),
+                                           j_);
+
+    if (!paymentParts)
+    {
+        XRPL_ASSERT_PARTS(
+            paymentParts.error(),
+            "ripple::LoanPay::doApply",
+            "payment error is an error");
+        return paymentParts.error();
+    }
+
+    // If the payment computation completed without error, the loanSle object
+    // has been modified.
+    view.update(loanSle);
+
+    XRPL_ASSERT_PARTS(
+        // It is possible to pay 0 principal
+        paymentParts->principalPaid >= 0,
+        "ripple::LoanPay::doApply",
+        "valid principal paid");
+    XRPL_ASSERT_PARTS(
+        // It is possible to pay 0 interest
+        paymentParts->interestPaid >= 0,
+        "ripple::LoanPay::doApply",
+        "valid interest paid");
+    XRPL_ASSERT_PARTS(
+        // It should not be possible to pay 0 total
+        paymentParts->principalPaid + paymentParts->interestPaid > 0,
+        "ripple::LoanPay::doApply",
+        "valid principal paid");
+    XRPL_ASSERT_PARTS(
+        paymentParts->feePaid >= 0,
+        "ripple::LoanPay::doApply",
+        "valid fee paid");
+
+    if (paymentParts->principalPaid < 0 || paymentParts->interestPaid < 0 ||
+        paymentParts->feePaid < 0)
+    {
+        // LCOV_EXCL_START
+        JLOG(j_.fatal()) << "Loan payment computation returned invalid values.";
+        return tecLIMIT_EXCEEDED;
+        // LCOV_EXCL_STOP
+    }
+
+    //------------------------------------------------------
+    // LoanBroker object state changes
+    view.update(brokerSle);
+
+    auto assetsAvailableProxy = vaultSle->at(sfAssetsAvailable);
+    // The vault may be at a different scale than the loan. Reduce rounding
+    // errors during the payment by rounding some of the values to that scale.
+    auto const vaultScale = assetsAvailableProxy->value().exponent();
+
+    auto const totalPaidToVaultRaw =
+        paymentParts->principalPaid + paymentParts->interestPaid;
+    auto const totalPaidToVaultRounded =
+        roundToAsset(asset, totalPaidToVaultRaw, vaultScale, Number::downward);
+    auto const totalPaidToVaultForDebt =
```

Comment:
This line of code might require commenting, as the logic is slightly inverse from the spec. What do you think about the following comment?


```suggestion
    // Account for value changes when reducing the broker's debt:
    // - Positive value change (from full/late/overpayments): Subtract from the
    //   amount credited toward debt to avoid over-reducing the debt.
    // - Negative value change (from full/overpayments): Add to the amount
    // credited toward debt,effectively increasing the debt reduction.
    auto const totalPaidToVaultForDebt =
        totalPaidToVaultRaw - paymentParts->valueChange;
```